packages/backend/src/apps/m365-excel/common/workbook-session/index.ts

```diff
@@ -0,0 +1,306 @@
+import type { IGlobalVariable } from '@plumber/types'
+
+import type { AxiosRequestConfig, AxiosResponse } from 'axios'
+
+import {
+  getM365TenantInfo,
+  type M365TenantInfo,
+} from '@/config/app-env-vars/m365'
+import HttpError from '@/errors/http'
+import logger from '@/helpers/logger'
+
+import { tryParseGraphApiError } from '../parse-graph-api-error'
+
+import {
+  clearSessionIdFromRedis,
+  getSessionIdFromRedis,
+  runWithLock,
+  setSessionIdInRedis,
+} from './redis'
+
+async function invalidateSessionId(
+  tenant: M365TenantInfo,
+  fileId: string,
+  badSessionId: string,
+): Promise<void> {
+  await runWithLock(tenant, fileId, async (signal) => {
+    // Nothing to do if another worker has already switched our fleet to a
+    // different session.
+    const sessionIdInRedis = await getSessionIdFromRedis(tenant, fileId)
+    if (sessionIdInRedis !== badSessionId) {
+      return
+    }
+
+    if (signal.aborted) {
+      throw new Error(`Redis lock error: ${signal.reason}`)
+    }
+
+    await clearSessionIdFromRedis(tenant, fileId)
+  })
+}
+
+async function refreshSessionId(
+  tenant: M365TenantInfo,
+  fileId: string,
+  $: IGlobalVariable,
+): Promise<string> {
+  return await runWithLock(tenant, fileId, async (signal) => {
+    // It's possible for multiple workers - or even multiple calls by the same
+    // worker - to await this. When this happens, the 1st caller will have
+    // refreshed the session and stored its ID in redis, so subsequent callers
+    // should use that stored session.
+    //
+    // We perform a check here to ensure that these subsequent callers use the
+    // refreshed session instead of requesting for a new one.
+    const sessionIdInRedis = await getSessionIdFromRedis(tenant, fileId)
+
+    if (sessionIdInRedis) {
+      return sessionIdInRedis
+    }
+
+    const createSessionResponse = await $.http.post<{ id: string }>(
+      `/v1.0/sites/${tenant.sharePointSiteId}/drive/items/${fileId}/workbook/createSession`,
+      {
+        persistChanges: true,
+      },
+    )
+    const sessionId = createSessionResponse.data.id
+
+    if (signal.aborted) {
+      throw new Error(`Redis lock error: ${signal.reason}`)
+    }
+
+    await setSessionIdInRedis(tenant, fileId, sessionId)
+    return sessionId
+  })
+}
+
+/**
```

Comment:
wah this is super complex to me, would most probably need to kindly request some time from you to walkthrough and clear some doubts

Reply:
ping me when you free, I can walk you through